src/common/file_location.c

```diff
@@ -82,120 +114,172 @@ gchar *dt_loc_get_home_dir(const gchar *user)
 #endif
 }
 
-gchar *dt_loc_init_generic(const char *value, const char *default_value)
+gchar *dt_loc_init_generic(const char *absolute_value, const char *application_directory, const char *default_value)
 {
-  const gchar *path = value ? value : default_value;
-  gchar *result = dt_util_fix_path(path);
-  if(g_file_test(result, G_FILE_TEST_EXISTS) == FALSE) g_mkdir_with_parents(result, 0700);
+  gchar *result = NULL;
+  gchar *path = NULL;
+  fprintf(stderr, "\tabsolute value: %s,\n\tapplication_directory: %s,\n\tdefault_value: %s\n", absolute_value, application_directory, default_value);
+  
+  if(absolute_value)
+  {
+    fprintf(stderr, "* absolute value. \n");
+    // the only adjustment the absolute path needs is transforming the possible tilde '~' to an absolute path
+    path = dt_util_fix_path(absolute_value);
+  }
+  else
+  {
+    // the default_value could be absolute or relative. we decide upon presence of the application_directory.
+    if(application_directory)
+    {   
+      // default_value is relative.
+      // combine basename (application_directory) and relative path (default_value).
+      gchar complete_path[PATH_MAX] = { 0 };
+#if defined(__APPLE__)
+      // fprintf(stderr, "* checking for bundle\n");
+      char *bundle_path = dt_osx_get_bundle_res_path();
+      if(bundle_path)
+      {
+        /// bundle detected ...
+        fprintf(stderr, "(bundle detected)\n");
+
+        // on a mac inside the bundle the executables are in <bundleroot>/Contents/MacOS/
+        // all other directories are a subdirectory of <bundleroot>/Contents/Resources:
+        // <bundleroot>/Contents/Resources/etc
+        // <bundleroot>/Contents/Resources/lib
+        // <bundleroot>/Contents/Resources/share
+        // so the relative path from the binary directory to the other directories differs to the non-bundle version by
+        // ../etc -> ../Resources/etc, 
+        // ../lib -> ../Resources/lib, 
+        // ../share -> ../Resources/share, 
+        // So we have to modify the relative default value
+
+        // +2: removes the two dots '..'
+        g_snprintf(complete_path, sizeof(complete_path), "%s/../Resources%s", application_directory, default_value + 2);
```

Comment:
Not very pretty solution since there is an assumption about how default_value is structured. E.g. it won't work if CMAKE_BINARY_DIR is something like "games/bin". The problem is (as I already said) that you define all paths relative to executable location, while it doesn't really matter. What you need is paths relative to prefix (e.g. /usr/local), then you either:
1. use value passed from command line if any;
2. if it's a Mac bundle, use dt_osx_get_bundle_res_path() as prefix (you now ignore what path it actually returns);
3. deduce prefix from executable location.

Reply:
I agree that this is not very pretty but compromises have to be made. I hope the implemented compromises are the best choice, though.

Whether my approach to set up everything relative to the bin directory (in which the executable darktable lives) can be debated. Changing the relative resolution technique to another one in a second step should be pretty easy because the big work in this PR is to enable relative resolution at all. We can happily discuss the ups and downs later on.

I don't understand why it won't work when CMAKE_BINARY_DIR is /games/bin (did you miss the leading slash?). It was explicitly designed to work with this location, that is why I came to the conclusion that I have to specify everything relative to the binary dir.

What I can see as a next version if necessary is #3. I detect the base dir of all darktable directories by the executable and then carry around ${BINDIR}/.. instead of saying ${BASEDIR}=${BINDIR}/.. and then using BASEDIR instead. However I really have to dig into that again to understand the complexity and how hard it is to change since I left these kind of thoughts end of May, when I thought it is finished.

Reply:
Let's say BINDIR is /usr/games/bin and LIBDIR is /usr/lib/. Then DARKTABLE_LIBDIR is ../../lib/darktable. With macOS bundle you will get in this line path like /Applications/darktable.app/Contents/MacOS/../Resources/../lib/darktable or /Applications/darktable.app/Contents/lib/darktable which is incorrect. But it doesn't matter much, since such setup is not supported by bundle creation script anyway.
So I agree that it makes sense to merge this PR as is, and maybe change approach as described in these comments later in a separate PR.